File: packages/typescript/src/services/message.ts
// packages/typescript/src/services/message.ts
import type { RpcClient } from "../transport/rpc.js";
import type { CapabilityGate } from "../capabilities.js";

export interface MessageService {
  send(params: MessageSendParams): Promise<MessageRecord>;
  inbox(params: MessageInboxParams): Promise<MessageInboxResult>;
  list(params?: MessageListParams): Promise<MessageListResult>;
}

export interface MessageRecord {
  id: string;
  from_agent_id: string;
  to_agent_id: string;
  tree_root_id: string;
  body: unknown;
  sent_at: string;
  delivered_at: string | null;
  read_at: string | null;
}

export interface MessageSendParams {
  to_agent_id: string;
  body: unknown;
}

export interface MessageInboxParams {
  agent_id: string;
}

export interface MessageInboxResult {
  messages: MessageRecord[];
}

export interface MessageListParams {
  agent_id?: string;
  tree_root_id?: string;
  cursor?: string;
  limit?: number;
}

export interface MessageListResult {
  items: MessageRecord[];
  cursor: string | null;
}

export function createMessageService(rpc: RpcClient, gate: CapabilityGate): MessageService {
  return {
    async send(params) {
      gate.require("message.send");
      return rpc.call<MessageRecord>("message.send", params);
    },
    async inbox(params) {
      gate.require("message.inbox");
      return rpc.call<MessageInboxResult>("message.inbox", params);
    },
    async list(params = {}) {
      gate.require("message.list");
      return rpc.call<MessageListResult>("message.list", params);
    },
  };
}

File: packages/python/helix_sdk/services/message.py
# packages/python/helix_sdk/services/message.py
from __future__ import annotations

from typing import Any

from ..capabilities import CapabilityGate
from ..transport import RpcClient


class MessageService:
    def __init__(self, rpc: RpcClient, gate: CapabilityGate) -> None:
        self._rpc = rpc
        self._gate = gate

    async def send(self, to_agent_id: str, body: Any) -> dict[str, Any]:
        self._gate.require("message.send")
        return await self._rpc.call("message.send", {"to_agent_id": to_agent_id, "body": body})

    async def inbox(self, agent_id: str) -> dict[str, Any]:
        self._gate.require("message.inbox")
        return await self._rpc.call("message.inbox", {"agent_id": agent_id})

    async def list(self, **params: Any) -> dict[str, Any]:
        self._gate.require("message.list")
        return await self._rpc.call("message.list", params)
